### ISSUE TITLE
really confusing / noisy haddocks for typeclass + data family heavy code 

### ISSUE THREAD

**Issue by cartazio** [haskell/haddock#300]
when i generate the haddocks for 
https://github.com/wellposed/numerical-core/blob/master/src/Numerical/Array/Layout/Base.lhs

It may very well be that the way its rendering is sensible,
but one concern I have is that the comments are shifted FAR to the right
for a given data family instance, and thats a bit unsatisfactory

(edit: i've now made a copy of the actual haddocks online) http://www.wellposed.com/examples-cached/Numerical-Array-Layout-Base.html should link to the relevant html

using ghc 7.8.2 haddock, heres some screen caps 

![screen shot 2014-05-09 at 6 44 43 pm](https://cloud.githubusercontent.com/assets/4290/2934193/cc2d7634-d7cb-11e3-86b3-49ea02ad91bc.png)
![screen shot 2014-05-09 at 6 44 33 pm](https://cloud.githubusercontent.com/assets/4290/2934194/cc3968fe-d7cb-11e3-8f68-44b75e881abd.png)

**Comment by tomsmalley** [comment]
Is there any movement on this? It's particularly noticeable in the servant-server docs: http://hackage.haskell.org/package/servant-server-0.10/docs/Servant-Server.html.

I think it would be good to wrap constraints in some `inline-block` `span`s and allow wrapping in the `td`, thereby breaking very long lines on the commas and `=>`.

**Comment by Fuuzetsu** [comment]
No movement.

**Comment by cartazio** [comment]
Bump: this issue is probably still biting code I write

**Comment by cartazio** [comment]
One direction might be to render type fam vs type clsss vs data fam instances in distinct groups?

There’s also the funny issue that if I define a single new Data fam constructor and export it, the docs in that module show alll visible constructors/ instances. So another collapsible axis is perhaps locally defined vs everything else that reaches that module?

**Comment by cartazio** [comment]
@harpocrates any thoughts on this one?  It was a bad enough leakage in some of my code that I redid how I wrote an api or two to avoid the extra noise in the docs! (Like an entire copy of vector instances being listed for everything local plus the ones coming from vector ...)

**Comment by harpocrates** [comment]
I agree this is gross. The main blocker here is just figuring out a nice alternate design. (I've contemplated this issue a fair bit and I will get around to it eventually - but don't hold your breath. 😄)

Note that the latest version of Haddock at least hides the actual data definitions (as in https://hackage.haskell.org/package/vector-0.12.0.1/docs/Data-Vector-Unboxed.html#t:Vector). 

Personally, I'd like to redo what gets shown once you expand a particular `data instance`. In particular, it would be nice to have the expanded output for any one `data instance` look as much as possible like what you would get for a plain old `data` definition. That way, we could properly handle things like documentation, fixity, source links, etc.

**Comment by cartazio** [comment]
I’d go further and argue that you need the nested or flat version of the
following


Data fam def
Instances drop down
   Instance of Data fam locally defined drop down
  Instances of dsta Fam inherited from elsewheee drop down

Similar for type fams and and type classes. Or something like that.

Could be 3 level nested. Or just all 4-6 singly nested in the alll exist
cases, each a different plus minus show hide toggle.

**Comment by cartazio** [comment]
Either way, sounds like we agree.

It might be as simple as doing a demand driven generate  multiple things
using the current instance template

**Comment by harpocrates** [comment]
@cartazio I've started to look into data families again. I have something preliminary implemented, but I'd really like your input, since you seem to care particularly about this feature.

Do you have some sufficiently complex data family code on hand? Once I have something I'm happy with, I'll open a PR with some screenshots and everyone can discuss what direction they want the design to go in.

**Comment by harpocrates** [comment]
On the data-family front, here's a WIP. I've reused the regular constructor format, and made sure docs and fixities make their way to the right places.

Source:

```haskell
{-# LANGUAGE TypeFamilies, GADTs, FlexibleInstances #-}
module DataFamily where

-- | some data family
data family Foo a
infixl 3 `Foo`

-- | some type family
type family Fam a :: *

-- | one type instance
type instance Fam (Foo Int) = Int

-- | another type instance
type instance Fam (Foo [a]) = a

-- | A newtype instance
newtype instance Foo [a] = Foo_List [a]

-- | A plain ADT instance
data instance Foo Int
  = Foo_Int Int -- ^ Int
  | Foo_Zero    -- ^ Zero
infixr 5 `Foo_Int`

-- | A GADT like instance
data instance Foo (a, b) where
  Foo_Eq :: a         -- ^ lhs
         -> a         -- ^ rhs
         -> Foo (a,a) -- ^ output
  Foo_MaybeEq :: a -> b -> Foo (a,b)
infixl 2 `Foo_Eq`

instance Eq (Foo Int) where
  Foo_Zero == Foo_Zero = True
  Foo_Zero == Foo_Int n = 0 == n
  Foo_Int n == Foo_Zero = n == 0
  Foo_Int n == Foo_Int m = n == m

```

Unexpanded:

![screenshot_2018-07-03 datafamily 1](https://user-images.githubusercontent.com/10766081/42254470-c81752d2-7f14-11e8-85ab-72878cba3901.png)

Expanded:

![screenshot_2018-07-03 datafamily](https://user-images.githubusercontent.com/10766081/42254480-d09b471a-7f14-11e8-92c4-40686e8ff77d.png)

**Comment by cartazio** [comment]
This looks like a clear improvement,  
the reference code / offending docs can be found http://www.wellposed.com.s3.amazonaws.com/numerical-0.1-prerelease-docs/Numerical-Array-Layout-Base.html#t:Format

one issue i have with the current way the instances stuff interacts with data families is we blob together 
1) type class instances when the type appears as an arg
2) type/data family instances where the appears as an arg
3) instances of the data family itself!

It seems like we should (possibly split them up more)

the above haddocks were generated a while ago from my repo https://github.com/wellposed/numerical/
